Enable mach bootstrap sandbox extension on macOS

pvollan opened the pull request:
#### 86a68721d5a98b019ba47e7ca7dc4c6166992397
<pre>
Enable mach bootstrap sandbox extension on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=252995">https://bugs.webkit.org/show_bug.cgi?id=252995</a>
rdar://problem/106291955

Reviewed by Sihui Liu and Simon Fraser.

Also macOS has the Mach bootstrap sandbox extension. This patch also makes sure that we
don&apos;t block launchd when not all GPUP flags are set, since we then can have Mach
connections being opened to the Metal compiler service at any time, which would not
succeed with blocking of launchd.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:

Canonical link: <a href="https://commits.webkit.org/261478@main">https://commits.webkit.org/261478@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed01eb769c4f2b7c345333d72a423a874e5d7061

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111795 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20924 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/410 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3568 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/120502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22282 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/11996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117560 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16550 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99698 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/104812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/100271 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/13387 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/275 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/11526 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13887 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/11996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/101516 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19328 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52277 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31554 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7988 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15869 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/109556 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26958 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -967,7 +967,7 @@ WebPage::WebPage(PageIdentifier pageID, WebPageCreationParameters&& parameters)
     }
 #if !ENABLE(LAUNCHD_BLOCKING_IN_WEBCONTENT) && HAVE(MACH_BOOTSTRAP_EXTENSION)
-    if (!(experimentalSandbox || experimentalSandboxWithProbability))
+    if (!(experimentalSandbox || experimentalSandboxWithProbability) || !shouldBlockIOKit)
```
Might be better put `!shouldBlockIOKit` on the left (as left operand will be evaluated first).

IIUC, web process now consumes `machBootstrapHandle` if experimentalSandbox is not enabled **or** shouldBlockIOKit is not true, but UI process only sends `machBootstrapHandle` when experimentalSandbox is not enabled. Is this expected? According to code in `WebPageProxy::creationParameters`:
```
#if !ENABLE(LAUNCHD_BLOCKING_IN_WEBCONTENT) && HAVE(MACH_BOOTSTRAP_EXTENSION)
    if (!preferences().experimentalSandboxEnabled())
        parameters.machBootstrapHandle = SandboxExtension::createHandleForMachBootstrapExtension();
#endif
```

pvollan replied:
> Might be better put `!shouldBlockIOKit` on the left (as left operand will be evaluated first).
> 

Done.

> IIUC, web process now consumes `machBootstrapHandle` if experimentalSandbox is not enabled **or** shouldBlockIOKit is not true, but UI process only sends `machBootstrapHandle` when experimentalSandbox is not enabled. Is this expected? According to code in `WebPageProxy::creationParameters`:
> 
> ```
> #if !ENABLE(LAUNCHD_BLOCKING_IN_WEBCONTENT) && HAVE(MACH_BOOTSTRAP_EXTENSION)
>     if (!preferences().experimentalSandboxEnabled())
>         parameters.machBootstrapHandle = SandboxExtension::createHandleForMachBootstrapExtension();
> #endif
> ```

That is a great point. This is a bug. I have changed the code in the UI process to always create the extension, and let all the logic for consumption be in the WebContent process.

Thanks for catching!

szewai reviewed (commented):
(I have fairly limited knowledge about this code so) this commit message and title is a bit confusing to me. From what I can tell, this patch enables mach bootstrap extension on macOS and makes web process consume the extension when IOKit is not blocked. Is "not blocking IOKit" equivalent to "not all GPUP flags are enabled"?

Can we re-title it to "Enable mach bootstrap extension extension on macOS"?

szewai reviewed (approved):
lgtm

pvollan committed:
Enable mach bootstrap sandbox extension on macOS
https://bugs.webkit.org/show_bug.cgi?id=252995
rdar://problem/106291955

Reviewed by Sihui Liu and Simon Fraser.

Also macOS has the Mach bootstrap sandbox extension. This patch also makes sure that we
don't block launchd when not all GPUP flags are set, since we then can have Mach
connections being opened to the Metal compiler service at any time, which would not
succeed with blocking of launchd.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:

Canonical link: https://commits.webkit.org/261478@main
